KAFKA-14894: MetadataLoader must call finishSnapshot after loading a snapshot

cmccabe opened the pull request:
The MetadataLoader must call finishSnapshot after loading a snapshot. This function removes
whatever was in the old snapshot that is not in the new snapshot that was just loaded. While this
is not significant when the old snapshot was the empty snapshot, it is important to do when we are
loading a snapshot on top of an existing non-empty image.

In initializeNewPublishers, the newly installed publishers should be given a MetadataDelta based on
MetadataImage.EMPTY, reflecting the fact that they are seeing everything for the first time.

cmccabe committed:
KAFKA-14894: MetadataLoader must call finishSnapshot after loading a snapshot

mumrah reviewed (commented):
Do we need a finishSnapshot call in initializeNewPublishers as well?

cmccabe committed:
Fix first publish to use an empty image in the delta.

Fix first publish to use an empty image in the delta.

add ImageReWriterTest.testCloseInvokesFinishSnapshot

mumrah reviewed (approved):
Thanks @cmccabe, LGTM.

cmccabe commented:
Frustrating that there are so many unrelated test flakes in `:connect:mirror:integrationTest`. Anyway thanks, committing.
